Massively enhance the Installation documentation (#160)
iv-org/documentation: Installation.md

After installation take a look at the [Post-install steps](#post-install-configuration).

> Note: Any [PaaS](https://en.wikipedia.org/wiki/Platform_as_a_service) or [SaaS](https://en.wikipedia.org/wiki/Software_as_a_service) provider/software (Heroku, YunoHost, Repli...) are unsupported, use them at your own risk, they **WILL** cause problems with Invidious and might even suspend your account for "abuse" since Invidious is heavy, bandwidth intensive and technically a proxy (and most provider don't like them). If you use one and want to report an issue, please mention which one you use.

Reviewer: Also:
```suggestion
Note: Any [PaaS](https://en.wikipedia.org/wiki/Platform_as_a_service) or [SaaS](https://en.wikipedia.org/wiki/Software_as_a_service) provider/software (Heroku, YunoHost, Repli...) are unsupported, use them at your own risk, they **WILL** cause problems with Invidious and might even suspend your account for "abuse" since Invidious is heavy, bandwidth intensive and technically a proxy (and most providers don't like them). If you use one and want to report an issue, please mention which one you use.
```

> Note: Any [PaaS](https://en.wikipedia.org/wiki/Platform_as_a_service) or [SaaS](https://en.wikipedia.org/wiki/Software_as_a_service) provider/software (Heroku, YunoHost, Repli...) are unsupported, use them at your own risk, they **WILL** cause problems with Invidious and might even suspend your account for "abuse" since Invidious is heavy, bandwidth intensive and technically a proxy (and most provider don't like them). If you use one and want to report an issue, please mention which one you use.

Author: Done

> ---

Author: Done

---

## Automated installation

[Invidious-Updater](https://github.com/tmiland/Invidious-Updater) is a self-contained script that can automatically install and update Invidious.

---

## Docker

> The Invidious docker image is only [available on Quay](https://quay.io/repository/invidious/invidious) because, unlike Docker Hub, [Quay is open source](https://github.com/quay/quay/blob/master/LICENSE). This is reflected in the `docker-compose.yml` file used in this walkthrough.{.is-warning}
> The Invidious docker image is only [available on Quay](https://quay.io/repository/invidious/invidious) because, unlike Docker Hub, [Quay is Free and Open Source Software](https://github.com/quay/quay/blob/master/LICENSE). This is reflected in the `docker-compose.yml` file used in this walk-through.{.is-warning}

Ensure [Docker Engine](https://docs.docker.com/engine/install/) and [Docker Compose](https://docs.docker.com/compose/install/) are installed before beginning.

### Make directory

```bash
$ mkdir invidious
```
### Docker-compose method (production)

### Create Docker Compose file

```bash
$ cd invidious
```

```bash
$ nano docker-compose.yml
```

> Here is a working Compose setup:

Author: Done

After installation take a look at the [Post-install steps](#post-install-configuration).

> Note: Any [PaaS](https://en.wikipedia.org/wiki/Platform_as_a_service) or [SaaS](https://en.wikipedia.org/wiki/Software_as_a_service) provider/software (Heroku, YunoHost, Repli...) are unsupported, use them at your own risk, they **WILL** cause problems with Invidious. If you use one and want to report an issue, please mention which one you use.

Reviewer 2: On most of the PaaS you can even have your account suspended.

Author: @unixfox Oh? Really? It's *that bad*?

Reviewer 2: Some don't want to deal with proxies services at all.

Allowing bots that excessively crawl (Semrush, webmeup, etc.) will lead your instance to get blocked very fast. While not required, it is a good idea to consider using bot blocking software such as [Nginx Bad Bot Blocker](https://github.com/mitchellkrogza/nginx-ultimate-bad-bot-blocker).

---

> 

Reviewer 3: Those should be removed imho.

Author: It's good to separate sections in markdown, we should really keep them imo, there's really no downside in having better separation between sections.

    ports:
      - "127.0.0.1:3000:3000"
    environment:

>       INVIDIOUS_CONFIG: | # Follow the config/config.yml file from the main repository to know the available configuration options

Author: Done

    ports:
      - "127.0.0.1:3000:3000"
    environment:
      INVIDIOUS_CONFIG: | 

>       # Follow the config/config.yml file from the main repository to know the available configuration options

Reviewer 3: ```suggestion
        # Please read the following file for a comprehensive list of all available
        # configuration options and their associated syntax:
        # https://github.com/iv-org/invidious/blob/master/config/config.example.yml
```

Author: Done

  invidious:
```

Note: This compose is made for a true "production" setup, where Invidious is behind a reverse proxy. If you prefer to directly access Invidious, replace `127.0.0.1:3000:3000` with `3000:3000` under the `ports:` section.

> The environment variable `POSTGRES_USER` cannot be changed. The SQL config files that run the initial database migrations are hard-coded with the username `kemal`.
{.is-warning}

> ### Docker-compose method (development)

Author: Done

  invidious:
```

> Note: This compose is made for a true "production" setup, and is made to be used to have Invidious behind a reverse proxy, if you don't want that and prefer to directly access Invidious via its port, change `ports: - "127.0.0.1:3000:3000"` to `ports: - "3000:3000"`.

Reviewer 3: ```suggestion
Note: This compose is made for a true "production" setup, where Invidious is behind a reverse proxy. If you prefer to directly access Invidious, replace `127.0.0.1:3000:3000` with `3000:3000` under the `ports:` section.
```

> The environment variable `POSTGRES_USER` cannot be changed. The SQL config files that run the initial database migrations are hard-coded with the username `kemal`.
{.is-warning}

### Docker-compose method (development)

```bash
git clone https://github.com/iv-org/invidious.git
cd invidious

> docker-compose build

Reviewer 3: `docker-compose build` is not required. `up` already builds the containers.

Author: Done

  invidious:
```

Note: This compose is made for a true "production" setup, where Invidious is behind a reverse proxy. If you prefer to directly access Invidious, replace `127.0.0.1:3000:3000` with `3000:3000` under the `ports:` section.

> The environment variable `POSTGRES_USER` cannot be changed. The SQL config files that run the initial database migrations are hard-coded with the username `kemal`.
{.is-warning}

### Docker-compose method (development)

> **This method build a Docker image from source**

Reviewer 3: builds*

### Docker-compose method (development)

> **This method build an image from source**

Reviewer 3: "This builds a docker image from source"

Author: I wanted to keep the same "style" between this one and the sentence used in the production method.

I just changed "image" to "Docker image" for both so that it's more clear.

Follow the instructions for your distribution here: https://crystal-lang.org/install/

If you're in a hurry, here are one-liner commands for some common distributions:
* Arch linux `sudo pacman -S crystal shards`
* Debian/Ubuntu: `curl -fsSL https://crystal-lang.org/install.sh | sudo bash`

> * Fedora: `sudo brew update && sudo brew install crystal-lang`

Author: We can't guarantee that this works, and it's better to have users take time to properly deploy an instance instead of saying "If you're in a hurry do *this*"

> $ sudo pacman -S base-devel librsvg postgresql

Author: Done

Arch Linux
```bash
sudo pacman -S base-devel librsvg postgresql
```

Ubuntu or Debian
```bash
sudo apt install libssl-dev libxml2-dev libyaml-dev libgmp-dev libreadline-dev postgresql librsvg2-bin libsqlite3-dev zlib1g-dev libpcre3-dev libevent-dev
```

> RHEL based and RHEL-like system (RHEL, Fedora, AlmaLinux, RockyLinux...)

Reviewer 3: systems*

Author: Done

#### Set up Invidious

```bash
$ su - invidious
$ cd invidious

> $ shards build src/invidious.cr --release

Author: Doesn't work:

```
Error target src/invidious.cr was not found in shard.yml.
```

(ping @SamantazFox )

Reviewer 3: Yeah, we need to make some changes first in the repo for that to work. You should probably revert for now.

Author: Done

$ exit
```

#### Systemd service

```bash
$ cp /home/invidious/invidious/invidious.service /etc/systemd/system/invidious.service

> $ systemctl enable --now invidious.service

Author: User is root

## Update Invidious

#### Updating a Docker install
```bash
$ docker-compose pull && docker-compose up && docker image prune -f
```

#### Update a manual install
```bash
$ sudo - invidious
$ cd invidious
$ currentVersion=$(git rev-list --max-count=1 --abbrev-commit HEAD)
$ git pull

> $ for i in `git rev-list --reverse --abbrev-commit $currentVersion..HEAD` ; do file=./config/migrate-scripts/migrate-db-$i.sh ; [ -f $file ] && $file ; done

Reviewer 3: Also, I'm in favor of keeping things separate. Update instructions should be in their own file.

> $ for i in `git rev-list --reverse --abbrev-commit $currentVersion..HEAD` ; do file=./config/migrate-scripts/migrate-db-$i.sh ; [ -f $file ] && $file ; done

Author: Done

Feel free to move the update instructions during the refactor.

$ shards build src/invidious.cr --release
$ exit
$ systemctl restart invidious.service
```

## Usage:

```bash
$ ./invidious
```

#### Logrotate configuration

```bash
$ echo "/home/invidious/invidious/invidious.log {
rotate 4
weekly
notifempty
missingok
compress
minsize 1048576
}" | tee /etc/logrotate.d/invidious.logrotate
$ chmod 0644 /etc/logrotate.d/invidious.logrotate

> ```

Reviewer 3: Ideally, this, DB maintenance cronjobs and systemd services should be moved to a separate "Post-install configuration" file, to keep that out of the way. I'm also tempted to put the logrotate file in the main repo. What do you think?

Author: To be discussed, open an issue please.

#### Updating a Docker install
```bash
docker-compose pull && docker-compose up && docker image prune -f
```

#### Update a manual install
```bash
sudo - invidious
cd invidious
shards update && shards install && crystal build src/invidious.cr --release
exit
systemctl restart invidious.service
```

## Usage:

```bash
./invidious
```

#### Logrotate configuration

```bash
echo "/home/invidious/invidious/invidious.log {
rotate 4
weekly
notifempty
missingok
compress
minsize 1048576
}" | tee /etc/logrotate.d/invidious.logrotate

> chmod 0644 /etc/logrotate.d/invidious.logrotate

Reviewer 3: `tee` and `chmod` require `sudo`

Author: It's implied that the user is root at this point.
